network-mux/src/Network/Mux.hs

```diff
@@ -157,26 +208,68 @@ muxStart tracer peerid app bearer = do
                               cnt
 
         return $ case app of
-          MuxInitiatorApplication initiator ->
-            [ ( initiator peerid mpdId initiatorChannel >> mpsJobExit cnt
+          MuxInitiatorApplication _ _ ->
+            [ ( mpsInitiatorBrooder itbl mpdId app initiatorChannel
               , show mpdId ++ " Initiator" )]
-          MuxResponderApplication responder ->
-            [ ( responder peerid mpdId responderChannel >> mpsJobExit cnt
-              , show mpdId ++ " Responder" )]
-          MuxInitiatorAndResponderApplication initiator responder ->
-            [ ( initiator peerid mpdId initiatorChannel >> mpsJobExit cnt
+          MuxResponderApplication _ _ ->
+            [ (mpsResponderBrooder pmss rtbl mpdId app responderChannel
+            , show mpdId ++ " Responder" )]
+          MuxInitiatorAndResponderApplication _ _ _ _ ->
+            [ ( mpsInitiatorBrooder itbl mpdId app initiatorChannel
               , show mpdId ++ " Initiator" )
-            , (responder peerid mpdId responderChannel >> mpsJobExit cnt
+            , (mpsResponderBrooder pmss rtbl mpdId app responderChannel
               , show mpdId ++ " Responder" )
             ]
 
-    -- cnt represent the number of SDUs that are queued but not yet sent.  Job
-    -- threads will be prevented from exiting until all SDUs have been
-    -- transmitted unless an exception/error is encounter. In that case all
-    -- jobs will be cancelled directly.
-    mpsJobExit :: StrictTVar m Int -> m ()
-    mpsJobExit cnt = do
+    -- Start and restart responder side miniprotocols on demand.
+    mpsResponderBrooder
+      :: HasResponder appType ~ True
+      => PerMuxSharedState ptcl m
+      -> MiniProtocolResponderControlTable ptcl m b
+      -> ptcl
+      -> MuxApplication appType peerid ptcl m a b
+      -> Channel m
+      -> m ()
+    mpsResponderBrooder pmss (MiniProtocolResponderControlTable rtbl) mpdId rspApp channel = forever $ do
+        -- Wait for initiator until we startup
+        traceWith tracer $ MuxTraceBrooderResponderBrood (AppProtocolId mpdId)
+        atomically $ do
+            buf <- readTVar (ingressQueue (dispatchTable pmss) (AppProtocolId mpdId) ModeResponder)
+            when (buf == BL.empty)
+                retry
+        -- Initiator is active, start the responder side
+        traceWith tracer $ MuxTraceBrooderResponderHatch (AppProtocolId mpdId)
+        b <- responderApplication rspApp peerid mpdId channel
+        -- Responder side exited, wait until we should hatch another responder
+        traceWith tracer $ MuxTraceBrooderResponderDone (AppProtocolId mpdId)
+        -- Propagate the miniprotocol result.
+        atomically $ putTMVar (rtbl ! AppProtocolId mpdId) b
```

Comment:
We're putting values inside a `TMVar` in a `forever` loop.  It should be explained, who reads them?  They will block if not taken in a timely manner.

Reply:
The values are consumed by the `MiniProtocolResponderControl` stm operation specified in the Mux application. I'll add a reference to it.